### PR TITLE
Path Distance Measurement Inconsistency

### PR DESCRIPTION
**Pull request by TheAlemazing**
close: https://github.com/fabricjs/fabric.js/issues/7327

### CODE REVIEW

File: src/util/path.js [hunk 1]
```diff
@@ -507,7 +507,9 @@
         p, nextLen, nextStep = 0.01, angleFinder = segInfo.angleFinder, lastPerc;
     // nextStep > 0.0001 covers 0.00015625 that 1/64th of 1/100
     // the path
-    while (tmpLen < distance && perc <= 1 && nextStep > 0.0001) {
+    
+    // floating point made perc <= 1 ignore some calculation, so perc < 1.1 do the trick
+    while (tmpLen < distance && perc < 1.1 && nextStep > 0.0001) {
```
Comment (asturur) [inline src/util/path.js:512]: why did they even bother adding floats to JS.

Comment (asturur) [reply src/util/path.js:512]: You left some trailing spaces, lint is failing.
This is good to go probably.
Can we look into reducing 1.1 to minimum possible? 1.01? 1.1 would be 10% of tolerance seems high.
I would also like to see a test that fails with 1 and pass with 1.0x

**Comment by TheAlemazing** [comment]
I will add the test soon.

File: src/util/path.js (hunk 1 above)
Comment (TheAlemazing) [reply src/util/path.js:511]: I think the correct should be
```
while (tmplen < distance && perc < (perc + nextstep) && nextstep > 0.0001)
```

File: src/util/path.js [hunk 2]
```diff
@@ -507,7 +507,7 @@
         p, nextLen, nextStep = 0.01, angleFinder = segInfo.angleFinder, lastPerc;
     // nextStep > 0.0001 covers 0.00015625 that 1/64th of 1/100
     // the path
-    while (tmpLen < distance && perc <= 1 && nextStep > 0.0001) {
+    while (tmpLen < distance && perc < (perc + nextStep) && nextStep > 0.0001) {
```
Comment (asturur) [inline src/util/path.js:510]: mmm this smells like infinite loop

Comment (asturur) [reply src/util/path.js:510]: can be this just < ( 1 + nextStep ), would this work?

Comment (TheAlemazing) [reply src/util/path.js:510]: yeah. I think 1 + nextStep is the correct one.

Comment (TheAlemazing) [reply src/util/path.js]: I tried, 1 + nextStep does not work. I think we can remove the percentage check as `tmpLen < distance` is enough.

Comment (TheAlemazing) [reply src/util/path.js]: I think we can remove the percentage check. Only `tmpLen < distance && nextStep > 0.0001` are enough

**Comment by TheAlemazing** [comment]
Some animate ezgif com-gif-maker

File: src/util/path.js [hunk 3]
```diff
@@ -507,15 +507,15 @@
...
+    while (tmpLen < distance && nextStep > 0.0001) {
       p = iterator(perc);
       lastPerc = perc;
       nextLen = calcLineLength(tempP.x, tempP.y, p.x, p.y);
       // compare tmpLen each cycle with distance, decide next perc to test.
       if ((nextLen + tmpLen) > distance) {
         // we discard this step and we make smaller steps.
-        nextStep /= 2;
         perc -= nextStep;
+        nextStep /= 2;
```
Comment (asturur) [inline src/util/path.js:518]: This could have been the bug.

**Comment by asturur** [comment]
@melchiar since you are the highest text on a path stakeholder here, do you want to have a look at this code?
I think is fine believe it will work fine.

File: src/util/path.js (hunk 3 above)
Comment (TheAlemazing) [reply src/util/path.js:518]: 1. Nope, it's not. I changed the line order to match the comment.
2. I think a bug relate to perc calculation, at the end of the path there are calculation steps that made perc greater than 1 (eg 1.00000000007) so the position calculation ignored. ( as we can see the text is not moving )

Comment (TheAlemazing) [reply src/util/path.js:518]: 3. We do not need to care about perc >= 1 because Distance and the tmpLen cover that.

**Comment by melchiar** [comment]
> @melchiar since you are the highest text on a path stakeholder here, do you want to have a look at this code? I think is fine believe it will work fine.

Sorry, been crazy busy and haven't had time to test yet. I will try to take a look on Monday

**Comment by melchiar** [comment]
Looks good to me!

**Review by asturur — APPROVED** [review]

**Comment by asturur** [comment]
You can formally approve next time!

**Pull request merged**